oap-server/server-query-plugin/logql-plugin/src/main/java/org/apache/skywalking/oap/query/logql/handler/LogQLApiHandler.java

```diff
@@ -0,0 +1,233 @@
+
+    @Get
+    @Path("/loki/api/v1/labels")
+    public HttpResponse labels(
+        @Param("start") Long start,
+        @Param("end") Long end) throws IOException {
+        LabelsQueryRsp labelsQueryRsp = new LabelsQueryRsp();
+        labelsQueryRsp.setStatus(ResultStatus.SUCCESS);
+
+        Duration duration = DurationUtils.timestamp2Duration(nano2Millis(start), nano2Millis(end));
+        tagAutoCompleteQueryService.queryTagAutocompleteKeys(TagType.LOG, duration)
+                                   .forEach(tag -> labelsQueryRsp.getData().add(tag));
+
+        return successResponse(labelsQueryRsp);
+    }
+
+    @Get
+    @Path("/loki/api/v1/label/{label_name}/values")
+    public HttpResponse labelValues(
+        @Param("label_name") String labelName,
+        @Param("start") Long start,
+        @Param("end") Long end) throws IOException {
+        LabelValuesQueryRsp response = new LabelValuesQueryRsp();
+        response.setStatus(ResultStatus.SUCCESS);
+
+        Duration duration = DurationUtils.timestamp2Duration(nano2Millis(start), nano2Millis(end));
+        tagAutoCompleteQueryService.queryTagAutocompleteValues(TagType.LOG, labelName, duration)
+                                   .forEach(value -> response.getData().add(value));
+
+        return successResponse(response);
+    }
+
+    @Get
+    @Path("/loki/api/v1/query_range")
+    public HttpResponse rangeQuery(
+        @Param("start") Long start,
+        @Param("end") Long end,
+        @Param("query") String query,
+        @Param("limit") Integer limit,
+        @Param("direction") LogDirection direction) throws IOException {
+        LogRangeQueryRsp logRangeQueryRsp = new LogRangeQueryRsp();
+        logRangeQueryRsp.setStatus(ResultStatus.SUCCESS);
+
+        LogQLLexer lexer = new LogQLLexer(CharStreams.fromString(query));
+        lexer.addErrorListener(new ParseErrorListener());
+        LogQLParser parser = new LogQLParser(new CommonTokenStream(lexer));
+        parser.addErrorListener(new ParseErrorListener());
+        ParseTree tree;
+        try {
+            tree = parser.root();
+        } catch (ParseCancellationException e) {
+            return badResponse(e.getMessage());
+        }
+
+        LogQLExprVisitor visitor = new LogQLExprVisitor();
+        LogQLParseResult parseResult = visitor.visit(tree);
+        Map<String, String> labelMap = parseResult.getLabelMap();
+
+        String serviceId = labelMap.containsKey(LabelName.SERVICE.getLabel()) ?
+            IDManager.ServiceID.buildId(labelMap.get(LabelName.SERVICE.getLabel()), true) : null;
+        String serviceInstanceId = labelMap.containsKey(LabelName.SERVICE_INSTANCE.getLabel()) ?
+            IDManager.ServiceInstanceID.buildId(serviceId, labelMap.get(LabelName.SERVICE_INSTANCE.getLabel())) : null;
+        String endpointId = labelMap.containsKey(LabelName.ENDPOINT.getLabel()) ?
+            IDManager.EndpointID.buildId(serviceId, labelMap.get(LabelName.ENDPOINT.getLabel())) : null;
```

Comment:
If `serviceId` is null, build instanceId and endpoint id will get unexpected id `_blank`.

Reply:
Ok, thanks. I will add a check.